File: src/java/magma/app/compile/define/Definition.java
package magma.app.compile.define;

import magma.api.collect.Joiner;
import magma.api.collect.list.List;
import magma.api.option.Option;
import magma.api.option.Some;
import magma.api.text.Strings;
import magma.app.Compiler;
import magma.app.compile.type.Type;

public record Definition(
        List<String> annotations,
        List<String> modifiers,
        List<String> typeParams,
        Type type,
        String name
) implements MethodHeader, Parameter {
    public Type findType() {
        return this.type;
    }

    public String toAssignment() {
        return "\n\t\tthis." + this.name + " = " + this.name + ";";
    }

    @Override
    public String generate() {
        return this.generateWithAfterName("");
    }

    @Override
    public Option<Definition> asDefinition() {
        return new Some<Definition>(this);
    }

    @Override
    public String generateWithAfterName(String afterName) {
        var joinedTypeParams = this.joinTypeParams();
        var joinedModifiers = this.modifiers.query()
                .map((String value) -> value + " ")
                .collect(new Joiner(""))
                .orElse("");

        return joinedModifiers + this.type.generateBeforeName() + this.name + joinedTypeParams + afterName + this.generateType();
    }

    private String generateType() {
        if (this.type.isVar()) {
            return "";
        }

        return ": " + this.type.generate();
    }

    private String joinTypeParams() {
        return Compiler.joinTypeParams(this.typeParams);
    }

    @Override
    public boolean hasAnnotation(String annotation) {
        return this.annotations.contains(annotation);
    }

    @Override
    public MethodHeader removeModifier(String modifier) {
        return new Definition(this.annotations, this.modifiers.removeValue(modifier), this.typeParams, this.type, this.name);
    }

    public boolean isNamed(String name) {
        return Strings.equalsTo(this.name, name);
    }
}

File: src/node/magma/app/compile/define/Definition.ts
import { List } from "../../../../magma/api/collect/list/List";
import { Type } from "../../../../magma/api/Type";
import { Platform } from "../../../../magma/app/io/Platform";
import { Option } from "../../../../magma/api/option/Option";
import { Some } from "../../../../magma/api/option/Some";
import { Joiner } from "../../../../magma/api/collect/Joiner";
import { Strings } from "../../../../jvm/api/text/Strings";
export class Definition {
	annotations: List<string>;
	modifiers: List<string>;
	typeParams: List<string>;
	type: Type;
	name: string;
	constructor (annotations: List<string>, modifiers: List<string>, typeParams: List<string>, type: Type, name: string) {
		this.annotations = annotations;
		this.modifiers = modifiers;
		this.typeParams = typeParams;
		this.type = type;
		this.name = name;
	}
	generate(platform: Platform): string {
		return this.generateWithAfterName(platform, "");
	}
	asDefinition(): Option<Definition> {
		return new Some<Definition>(this);
	}
	generateWithAfterName(platform: Platform, afterName: string): string {
		let joinedTypeParams: string = this.joinTypeParams();
		let joinedModifiers: string = this.joinModifiers();
		if (Platform.Windows === platform) {
			return joinedModifiers + this.type.generateBeforeName() + this.type.generate() + " " + this.name + afterName;
		}
		return joinedModifiers + this.type.generateBeforeName() + this.name + joinedTypeParams + afterName + this.generateType();
	}
	joinModifiers(): string {
		return this.modifiers.query().map((value: string) => value + " ").collect(new Joiner("")).orElse("");
	}
	generateType(): string {
		if (this.type.isVar()) {
			return "";
		}
		return ": " + this.type.generate();
	}
	joinTypeParams(): string {
		return Joiner.joinOrEmpty(this.typeParams, ", ", "<", ">");
	}
	hasAnnotation(annotation: string): boolean {
		return this.annotations.contains(annotation, Strings.equalsTo);
	}
	removeModifier(modifier: string): Definition {
		return new Definition(this.annotations, this.modifiers.removeValue(modifier, Strings.equalsTo), this.typeParams, this.type, this.name);
	}
	addModifierLast(modifier: string): Definition {
		return new Definition(this.annotations, this.modifiers.addLast(modifier), this.typeParams, this.type, this.name);
	}
	generateWithDefinitions(platform: Platform, definitions: List<Definition>): string {
		let joinedDefinitions: string = definitions.query().map((definition: Definition) => definition.generate(platform)).collect(new Joiner(", ")).orElse("");
		return this.generateWithAfterName(platform, "(" + joinedDefinitions + ")");
	}
}
